apps/admin-portal/src/app/resources/overview/resource-types/recipes/models.ts

```diff
@@ -1,4 +1,93 @@
-import { RecipeBuilder } from '@craftsmans-ledger/shared-ui';
+import { DestroyRef, EnvironmentInjector, FactoryProvider, inject, runInInjectionContext } from '@angular/core';
+import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
+import { ControlContainer, FormBuilder, FormControl, FormGroup, Validators } from '@angular/forms';
+import { RecipeBuilder, RecipeDto, RecipeItemDto, RecipesService, serialize } from '@craftsmans-ledger/shared-ui';
+import { filter, iif, merge, of, pairwise, startWith, switchMap, tap } from 'rxjs';
 import { TEMP_RESOURCE_ID } from '../../../models';
+import { ResourceService } from '../../resource.service';
+import { TEMP_TECHNOLOGY_TREE } from '../technology-trees/models';
 
-export const TEMP_RECIPE = new RecipeBuilder().withId(TEMP_RESOURCE_ID).build();
+export type RecipeItemDtoForm = FormGroup<{
+    itemId: FormControl<string>;
+    quantity: FormControl<number>;
+}>;
+
+export const TEMP_RECIPE = new RecipeBuilder()
+    .withId(TEMP_RESOURCE_ID)
+    .withTechnologyTree(TEMP_TECHNOLOGY_TREE)
+    .build()
+    .toDto();
+
+export function addRecipeItem(injector: EnvironmentInjector, type: string, value?: RecipeItemDto) {
+    return runInInjectionContext(injector, () => {
```

Comment:
Passing the injector here allows us to inject the services and components into this function. Injecting things requires a injection context which normally is only present in a class constructor or when assigning values to attributes of a class annotated with Injectable, Component, Directive, etc.